dnstable/entry.c

```diff
@@ -84,27 +84,39 @@ fmt_rfc3339_time(ubuf *u, uint64_t v)
 {
 	struct tm gm, *r;
 	time_t tm = v;
-	char s[sizeof("4294967295-12-31T23:59:59Z")];
+	char ts[]="2023-12-31T23:59:59Z";
 
 	r = gmtime_r(&tm, &gm);
-	if ((r != NULL) && (strftime(s, sizeof(s), "%Y-%m-%dT%H:%M:%SZ", r) > 0))
-		ubuf_add_cstr(u, s);
+	assert(r != NULL);
+
+	my_uint64_to_str_padded(1900 + r->tm_year, 4, ts);
+	my_uint64_to_str_padded(1 + r->tm_mon, 2, ts + 5);
+	my_uint64_to_str_padded(r->tm_mday, 2, ts + 8);
+	my_uint64_to_str_padded(r->tm_hour, 2, ts + 11);
+	my_uint64_to_str_padded(r->tm_min, 2, ts + 14);
+	my_uint64_to_str_padded(r->tm_sec, 2, ts + 17);
+
+	ubuf_append_cstr(u, ts, sizeof(ts)-1);
 }
 
 static yajl_gen_status
 fmt_rfc3339_time_json(yajl_gen g, uint64_t v)
 {
 	struct tm gm, *r;
 	time_t tm = v;
-	int ret;
-	char ts[sizeof("4294967295-12-31T23:59:59Z")];
+	char ts[]="2023-12-31T23:59:59Z";
 
 	r = gmtime_r(&tm, &gm);
 	assert(r != NULL);
 
-	ret = strftime(ts, sizeof(ts), "%Y-%m-%dT%H:%M:%SZ", r);
-	assert(ret > 0);
-	return yajl_gen_string(g, (const unsigned char *)ts, ret);
+	my_uint64_to_str_padded(1900 + r->tm_year, 4, ts);
```

Comment:
This is the exact same logic used in fmt_rfc3339_time(). Since the code is a non-trivial duplicate, the timestamp generation routine should be isolated into its own function and called by both of the functions that depend on it.